File: code/R/ch07/07.03.R
## 7.3 回帰分析

### 7.3.1 回帰分析とは何か

### 7.3.2 線形単回帰分析

library(tidyverse)

my_data <- cars
tmp <- data.frame(speed = 21.5, dist = 67)
my_data %>% ggplot(aes(x = speed, y = dist)) +
  coord_cartesian(xlim = c(4, 25), ylim = c(0, 120)) +
  geom_point() +
  stat_smooth(formula = y ~ x, method = "lm") +
  geom_pointrange(data = tmp, aes(ymin = -9, ymax = dist),  linetype = "dotted") +
  geom_pointrange(data = tmp, aes(xmin =  0, xmax = speed), linetype = "dotted")

### 7.3.3 回帰分析の実践

#### 7.3.3.1 データの用意

library(caret)
library(tidyverse)
my_data <- cars

#### 7.3.3.2 訓練

my_model <- train(form = dist ~ speed, # モデル式（出力変数と入力変数の関係）
                  data = my_data,      # データ
                  method = "lm")       # 手法

coef(my_model$finalModel)
#> (Intercept)       speed
#> -17.579095    3.932409

#### 7.3.3.3 予測

tmp <- data.frame(speed = 21.5)
my_model %>% predict(tmp)
#>        1
#> 66.96769

#### 7.3.3.4 モデルの可視化

f <- function(x) { my_model %>% predict(data.frame(speed = x)) }

my_data %>%
  ggplot(aes(x = speed, y = dist,
             color = "data")) +
  geom_point() +
  stat_function(
    fun = f,
    mapping = aes(color = "model"))

File: code/py/ch07/07.03.py
## 7.3 回帰分析

### 7.3.1 回帰分析とは何か

### 7.3.2 線形単回帰分析

import seaborn as sns
import statsmodels.api as sm

my_data = sm.datasets.get_rdataset('cars', 'datasets').data
ax = sns.regplot(x='speed', y='dist', data=my_data)
ax.vlines(x=21.5, ymin=-5, ymax=67,   linestyles='dotted')
ax.hlines(y=67,   xmin=4,  xmax=21.5, linestyles='dotted')
ax.set_xlim(4, 25)
ax.set_ylim(-5, 125)

### 7.3.3 回帰分析の実践

#### 7.3.3.1 データの用意

import statsmodels.api as sm
my_data = sm.datasets.get_rdataset('cars', 'datasets').data
X, y = my_data[['speed']], my_data['dist']

#### 7.3.3.2 訓練

# モデルの指定
from sklearn.linear_model import LinearRegression
my_model = LinearRegression()

# 訓練（モデルをデータにフィットさせる．）
my_model.fit(X, y)

# まとめて実行してもよい．
# my_model = LinearRegression().fit(X, y)

my_model.intercept_, my_model.coef_
#> (-17.579094890510973,
#>  array([3.93240876]))

#### 7.3.3.3 予測

tmp = [[21.5]]
my_model.predict(tmp)
#> array([66.96769343])

#### 7.3.3.4 モデルの可視化

import numpy as np
import pandas as pd

tmp = pd.DataFrame({'speed': np.linspace(min(my_data.speed),
                                         max(my_data.speed),
                                         100)})
tmp['model'] = my_model.predict(tmp)

pd.concat([my_data, tmp]).plot(
    x='speed', style=['o', '-'])
